src/coreclr/jit/optimizer.cpp

```diff
@@ -4168,49 +4188,47 @@ void Compiler::optInvertWhileLoop(BasicBlock* block)
         return;
     }
 
-    // Find the loop termination test at the bottom of the loop.
-    Statement* condStmt = bTest->lastStmt();
-
-    // bTest must only contain only a jtrue with no other stmts, we will only clone
-    // the conditional, so any other statements will not get cloned
-    //  TODO-CQ: consider cloning the whole bTest block and inserting it after block.
+    // It has to be a forward jump. Defer this check until after all the cheap checks
+    // are done, since it iterates forward in the block list looking for bbJumpDest.
+    //  TODO-CQ: Check if we can also optimize the backwards jump as well.
     //
-    if (bTest->bbStmtList != condStmt)
+    if (!fgIsForwardBranch(block))
     {
         return;
     }
 
-    /* Get to the condition node from the statement tree */
+    // Find the loop termination test at the bottom of the loop.
+    Statement* condStmt = bTest->lastStmt();
 
-    GenTree* condTree = condStmt->GetRootNode();
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* const condTree = condStmt->GetRootNode();
     noway_assert(condTree->gtOper == GT_JTRUE);
-
-    condTree = condTree->AsOp()->gtOp1;
-
-    // The condTree has to be a RelOp comparison
-    //  TODO-CQ: Check if we can also optimize the backwards jump as well.
-    //
-    if (condTree->OperIsCompare() == false)
+    if (!condTree->AsOp()->gtOp1->OperIsCompare())
     {
         return;
     }
 
-    // It has to be a forward jump. Defer this check until after all the cheap checks
-    // are done, since it iterates forward in the block list looking for bbJumpDest.
-    //  TODO-CQ: Check if we can also optimize the backwards jump as well.
+    // Estimate the cost of cloning the entire test block.
     //
-    if (fgIsForwardBranch(block) == false)
-    {
-        return;
-    }
-
-    /* We call gtPrepareCost to measure the cost of duplicating this tree */
+    // Note: it would help throughput to compute the maximum cost
+    // first and early out for large bTest blocks, as we are doing two
+    // tree walks per tree. But because of this helper call scan, the
+    // maximum cost depends on the trees in the block.
+    //
```

Comment:
> Any idea what this costs in terms of throughput?

I can do some pin measurements.

> I wonder if we could have some sort of early out here if the cost gets so large we can't imagine being able to shrink it back down with the discounts we try to apply later, just to avoid wasting time analyzing things we won't ever optimize.

Yeah, I was wondering about something like that. Might be easy to spit out stats for estimated cost versus max cost across some spmi collection.